openedx/core/djangoapps/util/tests/test_ip.py

```diff
@@ -0,0 +1,194 @@
+        ({'HTTP_SOME_THING': '::6, ::5, ::4, ::3, ::2, ::1'}, 'Some-Thing', -2, '::2', None),
+    )
+    def test_get_by_one_header(self, add_meta, header_name, index, expected, warning_substr):
+        self.request.META.update(add_meta)
+
+        with warning_messages() as caught_warnings:
+            actual = ip.get_client_ip_by_one_header(self.request.META, header_name, index)
+
+        assert actual == expected
+
+        if warning_substr is None:
+            assert len(caught_warnings) == 0
+        else:
+            assert len(caught_warnings) == 1
+            assert warning_substr in caught_warnings[0]
+
+    @ddt.unpack
+    @ddt.data(
+        # No config
+        ({}, [], None, 0),
+        # Configured, but none of the headers are present
+        (
+            {},
+            [
+                {'name': 'CF-Connecting-IP', 'index': 0},
+                {'name': 'X-Forwarded-For', 'index': -2},
+                {'name': 'X-Real-IP', 'index': 0},
+            ],
+            None,
+            3,
+        ),
+        # One lookup failure before finding a usable header
+        (
+            {
+                'HTTP_X_FORWARDED_FOR': '1.1.1.1, 2.2.2.2, 3.3.3.3, 4.4.4.4'
+            },
+            [
+                {'name': 'CF-Connecting-IP', 'index': 0},
+                {'name': 'X-Forwarded-For', 'index': -2},
+                {'name': 'X-Real-IP', 'index': 0},
+            ],
+            '3.3.3.3',
+            1,
+        ),
+    )
+    def test_get_via_configured_headers(self, add_meta, cnf_headers, expected, warning_count):
+        self.request.META.update(add_meta)
+
+        with override_settings(**{'CLIENT_IP_HEADERS': cnf_headers}):
+            with warning_messages() as caught_warnings:
+                actual = ip.get_client_ip_via_configured_headers(self.request.META)
+
+        assert actual == expected
+        assert len(caught_warnings) == warning_count
+
+    @ddt.unpack
+    @ddt.data(
+        # Nothing usable
+        ([], None),
+        (['any-old-junk'], None),
+        # Simple cases, private and public
+        (['::1'], '::1'),
+        (['1.2.3.4'], '1.2.3.4'),
+        # If we get a public IP, don't worry about junk farther on
+        (['junk', '1.2.3.4', '2606:4700::'], '2606:4700::'),
+        (['junk', '2606:4700::', '10.0.0.1'], '2606:4700::'),
+        # Walk left until first public IP
+        (['1.2.3.4', '5.6.7.8', '10.0.0.1', '127.0.0.1'], '5.6.7.8'),
+        # Or until there's junk, even if the best we can do is a private IP
+        (['1.2.3.4', 'XXXXXXX', '10.0.0.1', '127.0.0.1'], '10.0.0.1'),
+    )
+    def test_conservative_walk(self, chain, expected):
+        assert ip.conservatively_pick_client_ip(chain) == expected
+
+    @ddt.unpack
+    @ddt.data(
+        # Publicly exposed server, perhaps
+        ({'REMOTE_ADDR': '4.3.2.1'}, '4.3.2.1'),
+        # ...maybe with someone trying to spoof their IP
```

Comment:
1. Maybe this could be one of several tests under `Walk left until first public IP, starting with REMOTE_ADDR`, to use a language common to the other tests.
2. Not sure that mentioning `spoofing` helps here. Maybe it belongs in a comment in this file or in the actual code that explains the rules of: a) first public IP found, or b) last private IP found, and c) stop if junk found before any potential IP.